File: java/src/com/db4o/dg2db4o/chapter8/RunEmbeddedAddClient.java

package com.db4o.dg2db4o.chapter8;

import com.db4o.*;
import com.db4o.messaging.MessageSender;
import java.util.Scanner;

public class RunEmbeddedAddClient implements Runnable {
    ObjectServer server;
    ObjectContainer aClient = null;
    private boolean stop = false;
    Scanner sc = new Scanner(System.in);
    
    public RunEmbeddedAddClient(ServerRegistry sr) {
        server = sr.getServer("myserver");
    }
    
    public void run() {
        synchronized(this){
            try {
                aClient = server.openClient();
                while (!stop) { 
                    System.out.print("\nADDCLIENT:Enter the developer's name and age (e.g. 'Tom 44'): ");
                    String name = sc.next();
                    if(name.equals("stop"))
                        stop = true;
                    else {
                        aClient.set(new Person(name, sc.nextInt()));
                        aClient.commit();
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                MessageSender messageSender = aClient.ext().configure()
                .getMessageSender();
                messageSender.send(new StopServer("\nADDCLIENT says stop!"));
                aClient.close();
            }
        }
    }
    
    
}

File: csharp/Chapter07/Chapter08/RunEmbeddedAddClient.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using com.db4o;
using com.db4o.messaging;

namespace com.db4o.dg2db4o.chapter8
{
    class RunEmbeddedAddClient 
    {
        private ObjectServer server;
        private ObjectContainer aClient;
        private bool stop = false;

        public RunEmbeddedAddClient(ServerRegistry sr)
        {
            server = sr.GetServer("myserver");
        }

        public void Run()
        {
            lock (this)
            {
                try
                {
                    Console.WriteLine("Starting add client...");
                    aClient = server.OpenClient();
                    while (!stop)
                    {
                        Console.WriteLine("\nADDCLIENT: Please enter a name: ");
                        String name = Console.ReadLine();
                        if (name.Equals("stop"))
                        {
                            stop = true;
                        }
                        else
                        {
                            Console.WriteLine("\nADDCLIENT: Please enter an age: ");
                            int age = Convert.ToInt32(Console.ReadLine());
                            aClient.Set(new Person(name, age));
                            aClient.Commit();
                        }
                    }
                }
                catch (ThreadInterruptedException tie)
                {
                    Console.WriteLine("Thread Error!" + tie);
                }
                finally
                {
                    MessageSender messageSender = aClient.Ext().Configure()
                        .GetMessageSender();
                    messageSender.Send(new StopServer("ADDCLIENT says stop!"));
                    aClient.Close();
                }
            }

        }
    }
}
